File: src/Cpp/09_Listy_wprowadzenie/Zad12.cpp
#include <cassert>
#include <string>
#include <vector>

//Otrzymujesz liste liczb, kierunek przesuniec (1 odpowiada przesunieciu
//w prawo, a 0 w lewo) oraz liczbę miejsc o jaka maja zostac przesuniete
//elementy. Przykladowo dla przesuwania w prawo pierwszy element trafia
//na miejsce drugiego, drugi trzeciego, a ostatni na miejsce pierwszego.
//Przesun elementy listy w danym kierunku.

//Zlozonosc czasowa O(n)
//Zlozonosc pamieciowa O(n)
void rotacjaV1(std::vector<int>& lista, const std::string& kierunek, int liczba)
{

    if (kierunek == "prawo") {
        for (int i = 0; i < liczba; i++) {
            lista.insert(lista.begin(), lista.back());
            lista.erase(std::prev(lista.end()));
        }
    }

    else {
        for (int i = 0; i < liczba; i++) {
            lista.push_back(lista.front());
            lista.erase(lista.begin());
        }
    }
}

void test1()
{

    std::vector<int> lista{ 5, 27, 6, 2, 1, 10, 8 };
    std::vector<int> wynik{ 6, 2, 1, 10, 8, 5, 27 };
    std::string kierunek = "lewo";
    int liczba = 2;

    rotacjaV1(lista, kierunek, liczba);
    assert(lista == wynik);
}

void test2()
{

    std::vector<int> lista{ 9, 9, 42, 47, 5, 6, 19, 7 };
    std::vector<int> wynik{ 6, 19, 7, 9, 9, 42, 47, 5 };
    std::string kierunek = "prawo";
    int liczba = 3;

    rotacjaV1(lista, kierunek, liczba);
    assert(lista == wynik);
}

int main()
{

    test1();
    test2();

    return 0;
}

File: src/JavaScript/09_Listy_wprowadzenie/Zad12.js
   rotacjaV1 = function(lista, kierunek, liczba) {
       if (kierunek === ("prawo")) {
           for (var i = 0; i < liczba; i++) {
               {
                   lista.splice(0, 0, lista[lista.length - 1]);
                   lista.splice(lista.length - 1, 1)[0];
               };
           }
       } else {
           for (var i = 0; i < liczba; i++) {
               {
                   (lista.push(lista[0]) > 0);
                   lista.splice(0, 1)[0];
               };
           }
       }
       return lista;
   };
   test1 = function() {
       var lista = [5, 27, 6, 2, 1, 10, 8]
       var kierunek = "lewo";
       var liczba = 2;
       var oczekiwane = [6, 2, 1, 10, 8, 5, 27]
       var wynik = rotacjaV1(lista, kierunek, liczba);
       if (!(wynik.length == oczekiwane.length)) {
           throw new Error(`Assertion error line 59: ${wynik.length} != ${oczekiwane.length}`);
       }
       for (var i = 0; i < wynik.length; i++) {
           if (!(wynik[i] == oczekiwane[i])) {
               throw new Error(`Assertion error line 59: ${wynik[i]} != ${oczekiwane[i]}`);
           }
       }
   };
   test2 = function() {
       var lista = [9, 9, 42, 47, 5, 6, 19, 7]
       var liczba = 3;
       var kierunek = "prawo";
       var oczekiwane = [6, 19, 7, 9, 9, 42, 47, 5]
       var wynik = rotacjaV1(lista, kierunek, liczba);
       if (!(wynik.length == oczekiwane.length)) {
           throw new Error(`Assertion error line 69: ${wynik.length} != ${oczekiwane.length}`);
       }
       for (var i = 0; i < wynik.length; i++) {
           if (!(wynik[i] == oczekiwane[i])) {
               throw new Error(`Assertion error line 69: ${wynik[i]} != ${oczekiwane[i]}`);
           }
       }
   };
   main = function(args) {
       test1();
       test2();
   };

   main(null);